### PR TITLE
Welcome ui

### PR DESCRIPTION
**Pull request by mattkrick**
everything is real good here except for submitting the 3rd part to the server. gotta sign off for a couple hours as I move into the new digs, but should have a fix tonight!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
     "start": "NODE_ENV=production node ./src/server/server.babel.js",
     "dev": "NODE_ENV=development node ./src/server/server.babel.js",
     "build": "rimraf build && concurrently \"npm run build:client\" \"npm run build:server\"",
-    "bs": "rimraf build && concurrently \"npm run build:client\" \"npm run build:server\" \"npm run start\"",
+    "bs": "npm run build && npm start",
```
Comment (jordanh) [inline package.json]: Channeling your brain – I can see how you were looking at this redundancy and it just bothered the heck out of you didn't it? 👅

File: src/server/graphql/models/CachedUser/cachedUserMutation.js [hunk 2]
```diff
@@ -48,23 +45,18 @@ export default {
...
-      const changes = await r.table('CachedUser').insert(newUserObj, {
+      const changes = await r.table('CachedUser').insert(newUser, {
         conflict: 'update',
         returnChanges: true
       });
       // Did we update an existing cached profile?
-      if (changes.replaced > 0) {
-        return newUserAndToken;
+      if (changes.replaced === 0) {
+        const emailWelcomed = await sendEmail('newUser', newUser);
```
Comment (jordanh) [inline src/server/graphql/models/CachedUser/cachedUserMutation.js]: This is a fun way to trigger the welcome email. I like it. 👍

File: src/universal/decorators/requireAuthAndRole/requireAuthAndRole.js [hunk 3]
```diff
@@ -1,43 +1,48 @@
...
-            return <ComposedComponent {...this.props} auth={auth} />;
-          } else if (role === undefined) {
-            // We were looking for any authenticated user only:
-            return <ComposedComponent {...this.props} auth={auth} />;
-          }
-        }
-        // TODO: redirect back from whence they came
-        cashay.store.dispatch(showError(unauthorized));
-        cashay.store.dispatch(push('/'));
+export default role => ComposedComponent => {
```
Comment (jordanh) [inline src/universal/decorators/requireAuthAndRole/requireAuthAndRole.js]: Very much cleaner, better! 👍

File: src/universal/modules/welcome/components/WelcomeWizardForms/Step1PreferredName.js [hunk 4]
```diff
@@ -6,9 +6,25 @@ import WelcomeContent from '../WelcomeContent/WelcomeContent';
 import WelcomeHeader from '../WelcomeHeader/WelcomeHeader';
 import WelcomeHeading from '../WelcomeHeading/WelcomeHeading';
 import WelcomeLayout from '../WelcomeLayout/WelcomeLayout';
+import {cashay} from 'cashay';
+import {nextPage} from 'universal/modules/welcome/ducks/welcomeDuck';
 
 const Step1PreferredName = props => {
-  const {handleSubmit, preferredName} = props;
+  const {handleSubmit, preferredName, dispatch} = props;
+  const onPreferredNameSubmit = data => {
```
Comment (jordanh) [inline src/universal/modules/welcome/components/WelcomeWizardForms/Step1PreferredName.js]: Yeah, I like this better. It's easy for somebody to grok if they only had a single file to look at.

File: src/universal/modules/welcome/components/WelcomeWizardForms/Step3InviteTeam.js [hunk 5]
```diff
@@ -51,61 +44,102 @@ export default class Step3InviteTeam extends Component {
...
     });
-  }
+  };
 
-  render() {
-    const {handleSubmit, invitees, inviteesRaw, submitting, teamName} = this.props;
+  const onInviteTeamSubmit = async data => {
+    const serverInvitees = data.invitees.map(invitee => {
+      // Remove label field:
+      const {label, ...inviteeForServer} = invitee;
```
Comment (jordanh) [inline src/universal/modules/welcome/components/WelcomeWizardForms/Step3InviteTeam.js]: Noice.

File: src/universal/modules/welcome/components/WelcomeWizardForms/Step3InviteTeam.js [hunk 6]
```diff
@@ -51,61 +44,102 @@ export default class Step3InviteTeam extends Component {
...
+        invitees: serverInvitees
+      }
+    };
 
-    const invitesFieldHasError = false; // TODO: wire this up for real
-    const helpText = invitesFieldHasError ?
-      // eslint-disable-next-line max-len
-      <span>Oops! Please make sure email addresses are valid <br />and separated by a single comma.</span> :
-      <span>You can paste multiple emails separated by a comma.<br />&nbsp;</span>;
+    const res = await cashay.mutate('inviteTeamMembers', options);
```
Comment (jordanh) [inline src/universal/modules/welcome/components/WelcomeWizardForms/Step3InviteTeam.js]: I mean, this reads to nicely. I don't think you should feel bad about the Promise interface here. It looks great!

**Comment by jordanh** [comment]
This all looks just excellent to me. I was going to start merging in your persist-fixes this morning but I see you've already done that. Imma going to fix some of your linting errors this morning before moving on to some of the other tasks on my task list.

**Comment by jordanh** [comment]
See: https://github.com/mattkrick/action/pull/3

**Comment by mattkrick** [comment]
Huh, looks like GraphQL changed their error handling on the server. All set now!
Put 2 TODOs in step3, but I figured this PR is big enough & wanted to get it merged ASAP before biting off some mo.

**Pull request merged**